{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Advanced Examples\n",
    "In this notebook, a number of advanced examples using BioCRNpyler will be presented. These are desiged to highlight a number of different built-in Mechanisms and Components and how they can be used in an object oriented fashion. For details on how to create your own Mechanisms, Components, and Mixtures, see the Developer Overview Notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example 1: Retroactivity and Loading\n",
    "Most of the default Mixtures in BioCRNpyler include transcription, translation, and degredation machinery such as RNAP (RNA Polymerase), Ribosomes, and RNAases. In the following example, we will illustrate loading effects due to competition over this machinery. For this example, we will use the following model of constituitive transcription and translation:\n",
    "\n",
    "$G_i + \\textrm{RNAP} \\leftrightarrow G_i:\\textrm{RNAP} \\rightarrow G_i + \\textrm{RNAP} + T_i$\n",
    "\n",
    "$T_i + \\textrm{Ribosome} \\leftrightarrow T_i:\\textrm{Ribosome} \\rightarrow T_i + \\textrm{Ribosome} + P_i$\n",
    "\n",
    "$T_i + \\textrm{RNAase} \\leftrightarrow T_i:\\textrm{RNAase} \\rightarrow \\textrm{RNAase}$\n",
    "\n",
    "Here $G_i$, $T_i$ and $P_i$ are gene, transcript, and protein $i$, respectively. In the example that follows, we will allow different RNA polymerases for different genes. Also, some genes may not be translated at all. By using orthogonal polymerases and/or loads without translation, we will see different kinds of loading effects.\n",
    "\n",
    "We will create 4 different DNA assemblies. The reference assembly, \"ref\", will have a RNAP promoter and an RBS. We will examine the output of this reporter as a function of the amount of various load assemblies.\n",
    "* The \"Load\" assembly will be identical to the \"ref\" assembly; this assembly will put load on all parts of transcription, RNA degredation, and translation for the ref assembly.\n",
    "* The \"TxLoad\" assembly will have an RNAP promoter, but no RBS, ensuring that only RNAP and RNAases experience loading, not ribosomes.\n",
    "* The \"T7Load\" assembly will have a T7 promoter and an RBS so there will be no loading on polymerases.\n",
    "* the \"T7TxLoad\" assmebly will have a T7 promoter and no RBS, so there will only be load on the RNAases.\n",
    "\n",
    "The creation of these assemblies highlights the flexible, object oriented nature of bioCRNpyler."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from biocrnpyler import *\n",
    "%matplotlib inline\n",
    "\n",
    "#Because we are lazy, all parameters will use the default \"param_name\"-->value key mapping.\n",
    "#Parameter warnings will be later suppressed in the Mixture constructor\n",
    "#For details on how parameter loading and defaulting works, see the Parameter notebook.\n",
    "kb, ku, ktx, ktl, kdeg = 100, 10, 3.0, 5.0, 2\n",
    "parameters = {\"kb\":kb, \"ku\":ku, \"ktx\":ktx, \"ktl\":ktl, \"kdeg\":kdeg}\n",
    "\n",
    "#A constituitively expressed reporter\n",
    "#By default the promoter 'P' will use the polymerase 'rnap'\n",
    "reference_assembly = DNAassembly(name = \"ref\", promoter = \"P\", rbs = \"BCD\")\n",
    "#A constiuitively expressed load (RNA and Protein)\n",
    "full_load_assembly = DNAassembly(name = \"Load\", promoter = \"P\", rbs = \"BCD\")\n",
    "#A constiutively transcribed (but not translated) load\n",
    "#By putting rbs = None, DNAassembly automatically knows not to include translation\n",
    "RNA_load_assembly = DNAassembly(name = \"TxLoad\", promoter = \"P\", rbs = None)\n",
    "\n",
    "#Load genes on orthogonal polymerases\n",
    "T7 = Protein(\"T7\") #Create a new protein (polymerase) called 'T7'\n",
    "\n",
    "#Create a custom promoter with a custom mechanism that uses T7 instead of RNAP\n",
    "#instantiate a new mechanism Transcription_MM with its own name and overwrote the default parameter rnap\n",
    "mechanism_txt7 = Transcription_MM(name = \"T7_transcription_mm\", rnap=T7)\n",
    "#Create an instance of a promoter with this mechanism for transcription\n",
    "T7P = Promoter(\"T7P\", mechanisms={\"transcription\":mechanism_txt7})\n",
    "#Create A load assembly with the custom T7 promoter\n",
    "T7_load_assembly = DNAassembly(name = \"T7Load\", promoter = T7P, rbs = \"BCD\")\n",
    "\n",
    "#Each new assembly requires its own promoter instance - so here I create another here\n",
    "T7P = Promoter(\"T7P\", mechanisms={\n",
    "    \"transcription\":Transcription_MM(name = \"T7_transcription_mm\", rnap=T7)})\n",
    "#A load assembly with the custom T7 promoter and no RBS\n",
    "T7RNA_load_assembly = DNAassembly(name = \"T7TxLoad\", promoter = T7P, rbs = None)\n",
    "\n",
    "#Add all the assemblies to a mixture\n",
    "components = [reference_assembly, full_load_assembly, T7_load_assembly, T7, RNA_load_assembly, T7RNA_load_assembly]\n",
    "myMixture = BasicExtract(name = \"txtl\", parameters = parameters, components = components, parameter_warnings = False)\n",
    "\n",
    "#Print the CRN\n",
    "myCRN = myMixture.compile_crn()\n",
    "print(\"\\n\"+repr(myCRN)+\"\\n\")\n",
    "\n",
    "\n",
    "#Simulate with BioSCRAPE if installed\n",
    "try:\n",
    "    print(\"Simulating\")\n",
    "    import numpy as np\n",
    "    import pylab as plt\n",
    "    timepoints = np.arange(0, 3, .01)\n",
    "    stochastic = True #Whether to use ODE models or Stochastic SSA models\n",
    "    plt.figure(figsize = (16, 8))\n",
    "    plt.subplot(221)\n",
    "    plt.title(\"Load on a RNAP Promoter\")\n",
    "    loads = [0, 1.0, 5., 10., 50, 100, 500, 1000]\n",
    "    for dna_Load in loads:\n",
    "        #print(\"Simulating for dna_Load=\", dna_Load)\n",
    "        x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "                   'dna_ref':5., 'dna_Load':dna_Load}\n",
    "\n",
    "        results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label = \"Load = \"+str(dna_Load))\n",
    "\n",
    "    plt.xlim(0, 5)\n",
    "    #plt.xlabel(\"time\")\n",
    "    plt.ylabel(\"Reference Protein\")\n",
    "    plt.legend()\n",
    "\n",
    "    plt.subplot(222)\n",
    "    plt.title(\"Load on a T7 Promotoer\")\n",
    "    for dna_Load in loads:\n",
    "        #print(\"Simulating for dna_T7Load=\", dna_Load)\n",
    "        x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "                   'dna_ref':5., 'dna_T7Load':dna_Load}\n",
    "        results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label=\"Load = \" + str(dna_Load))\n",
    "    plt.xlim(0, 5)\n",
    "    #plt.xlabel(\"time\")\n",
    "    plt.ylabel(\"Reference Protein\")\n",
    "    plt.legend()\n",
    "\n",
    "    plt.subplot(223)\n",
    "    plt.title(\"Load on a RNAP Promotoer, No RBS\")\n",
    "    for dna_Load in loads:\n",
    "        #print(\"Simulating for dna_TxLoad=\", dna_Load)\n",
    "        x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "                   'dna_ref':5., 'dna_TxLoad':dna_Load}\n",
    "        results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label=\"Load = \" + str(dna_Load))\n",
    "    plt.xlim(0, 5)\n",
    "    plt.xlabel(\"time\")\n",
    "    plt.ylabel(\"Reference Protein\")\n",
    "    plt.legend()\n",
    "\n",
    "    plt.subplot(224)\n",
    "    plt.title(\"Load on a T7 Promotoer, No RBS\")\n",
    "    for dna_Load in loads:\n",
    "        #print(\"Simulating for dna_T7TxLoad=\", dna_Load)\n",
    "        x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "                   'dna_ref':5., 'dna_T7TxLoad':dna_Load}\n",
    "        results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label=\"Load = \" + str(dna_Load))\n",
    "    plt.xlim(0, 5)\n",
    "    plt.xlabel(\"time\")\n",
    "    plt.ylabel(\"Reference Protein\")\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "except ModuleNotFoundError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example 2: Regulated Promoters\n",
    "In this example we will consider combinatoric promoter regulated with an activator and a repessor. The reaction schema used by the RegulatedPromoter class is:\n",
    "$P + n_i*R_i \\leftrightarrow P:nR_i$ which represents $n_i$ copies of the regulator $R_i$ binding to the promoter $P$ in a single step.\n",
    "$P:nR_i + rnap \\leftrightarrow P:nR:rnap \\rightarrow P:n_i*R_i + rnap + T$ which represents the binding of $rnap$ to the promoter-regulator complex and subsequent transcription to produce $T$. Mechalis Menten translation and mRNA degredation as described for a TX-TL extract are also included in this model.\n",
    "\n",
    "To have both an activator and a repressor, we will allow the promoter 'P_reg' to bind to a regulator 'activator' and a regulator 'repressor'. The RNAP binding rates and the transcription rates for the regulator-promoter complexes will be chosen to give rise to activation and repression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Species = complex_protein_RNAP_protein_4.0xactivator_dna_reporter, complex_protein_4.0xactivator_dna_reporter, rna_reporter, complex_protein_RNAase_rna_reporter, protein_reporter, complex_protein_Ribo_rna_reporter, complex_protein_RNAP_protein_4.0xrepressor_dna_reporter, dna_reporter, complex_protein_4.0xrepressor_dna_reporter, complex_protein_RNAP_dna_reporter, protein_activator, protein_repressor, protein_RNAP, protein_Ribo, protein_RNAase\n",
      "Reactions = [\n",
      "\tdna_reporter + protein_RNAP <--> complex_protein_RNAP_dna_reporter        massaction: k_f=100,\tk_r=100\n",
      "\tcomplex_protein_RNAP_dna_reporter --> dna_reporter + rna_reporter + protein_RNAP        massaction: k_f=self.k\n",
      "\t4.0*protein_activator + dna_reporter <--> complex_protein_4.0xactivator_dna_reporter        massaction: k_f=1000,\tk_r=5.0\n",
      "\tcomplex_protein_4.0xactivator_dna_reporter + protein_RNAP <--> complex_protein_RNAP_protein_4.0xactivator_dna_reporter        massaction: k_f=1000,\tk_r=10.0\n",
      "\tcomplex_protein_RNAP_protein_4.0xactivator_dna_reporter --> complex_protein_4.0xactivator_dna_reporter + rna_reporter + protein_RNAP        massaction: k_f=self.k\n",
      "\t4.0*protein_repressor + dna_reporter <--> complex_protein_4.0xrepressor_dna_reporter        massaction: k_f=1000,\tk_r=5.0\n",
      "\tcomplex_protein_4.0xrepressor_dna_reporter + protein_RNAP <--> complex_protein_RNAP_protein_4.0xrepressor_dna_reporter        massaction: k_f=10,\tk_r=1000.0\n",
      "\tcomplex_protein_RNAP_protein_4.0xrepressor_dna_reporter --> complex_protein_4.0xrepressor_dna_reporter + rna_reporter + protein_RNAP        massaction: k_f=self.k\n",
      "\trna_reporter + protein_Ribo <--> complex_protein_Ribo_rna_reporter        massaction: k_f=50,\tk_r=10.0\n",
      "\tcomplex_protein_Ribo_rna_reporter --> rna_reporter + protein_reporter + protein_Ribo        massaction: k_f=self.k\n",
      "\trna_reporter + protein_RNAase <--> complex_protein_RNAase_rna_reporter        massaction: k_f=500,\tk_r=1.0\n",
      "\tcomplex_protein_RNAase_rna_reporter --> protein_RNAase        massaction: k_f=self.k\n",
      "]\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'py_simulate_model'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-144396c33b69>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     67\u001b[0m     \u001b[0;31m#Simulate the CRN with no activator or repressor\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m     \u001b[0mx0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"protein_activator\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"protein_repressor\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"dna_reporter\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"protein_Ribo\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"protein_RNAP\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"protein_RNAase\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 69\u001b[0;31m     \u001b[0mR_const\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmyCRN\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimulate_with_bioscrape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstochastic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minitial_condition_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m     \u001b[0;31m#Simulate teh CRN with just repressor\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/sclamons/Downloads/biocrnpyler_local/biocrnpyler/chemical_reaction_network.py\u001b[0m in \u001b[0;36msimulate_with_bioscrape\u001b[0;34m(self, timepoints, initial_condition_dict, stochastic, return_dataframe, safe)\u001b[0m\n\u001b[1;32m    567\u001b[0m                                 \u001b[0mstochastic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_dataframe\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    568\u001b[0m                                 safe = True):\n\u001b[0;32m--> 569\u001b[0;31m         \u001b[0;32mfrom\u001b[0m \u001b[0mbioscrape\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimulator\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpy_simulate_model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    570\u001b[0m         \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_bioscrape_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    571\u001b[0m         \u001b[0mm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_species\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minitial_condition_dict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'py_simulate_model'"
     ]
    }
   ],
   "source": [
    "import biocrnpyler as bcp\n",
    "import numpy as np\n",
    "\n",
    "#All Necessary Parameters are passed in as a dictionary in the form: (mechanism, part_id, param_name):Val\n",
    "#These could also be loaded from a file, but are shown here to illustrate the required parameter keys.\n",
    "parameters = {\n",
    "    #Promoter Activator Binding Parameters. Note the part_id = [promoter_name]_[regulator_name]\n",
    "    ('one_step_cooperative_binding', 'P_reg_activator', 'kb'):1000, #Promoter - Activator Binding\n",
    "    ('one_step_cooperative_binding', \"P_reg_activator\", 'ku'):5.0, #Unbinding\n",
    "    ('one_step_cooperative_binding',\"P_reg_activator\", 'cooperativity'):4.0, #Cooperativity\n",
    "    \n",
    "    #Activated Promoter Transcription. Note the part_id = [promoter_name]_[regulator_name]\n",
    "    #These regulate RNAP binding to an activated promoter and transcription\n",
    "    ('transcription_mm', 'P_reg_activator', 'kb'): 1000, #Binding to RNAP\n",
    "    ('transcription_mm', \"P_reg_activator\", 'ku'): 10.0, #Unbinding\n",
    "    ('transcription_mm', 'P_reg_activator', \"ktx\"): 1.25, #Transcription Rate\n",
    "    \n",
    "    #Promoter Repressor Binding Parameters. Note the part_id = [promoter_name]_[regulator_name]\n",
    "    ('one_step_cooperative_binding', 'P_reg_repressor', 'kb'):1000,\n",
    "    ('one_step_cooperative_binding',\"P_reg_repressor\", 'ku'):5.0,\n",
    "    ('one_step_cooperative_binding',\"P_reg_repressor\", 'cooperativity'):4.0,\n",
    "    \n",
    "    #Repressed Promoter Transcription. Note the part_id = [promoter_name]_[regulator_name]\n",
    "    #These regulate RNAP binding to a repressed promoter and transcription\n",
    "    ('transcription_mm', 'P_reg_repressor', 'kb'): 10, #Binding to RNAP\n",
    "    ('transcription_mm', \"P_reg_repressor\", 'ku'): 1000.0, #Unbinding\n",
    "    ('transcription_mm', 'P_reg_repressor', \"ktx\"): .25, #Transcription Rate\n",
    "    \n",
    "    #Leak Parameters for transcription\n",
    "    #These regulate expression of an unbound promoter\n",
    "    ('transcription_mm', 'P_reg', \"kb\"): 100, #Binding to RNAP\n",
    "    ('transcription_mm', 'P_reg', \"ku\"): 100, #Unbinding\n",
    "    ('transcription_mm', 'P_reg', \"ktx\"): 1., #Transcription Rate\n",
    "              \n",
    "    #Translation Parameters:\n",
    "    ('translation_mm', 'BCD', 'kb'): 50, #Binding to Ribosome\n",
    "    ('translation_mm', 'BCD', 'ku'): 10.0, #Unbinding\n",
    "    ('translation_mm', 'BCD', 'ktl'): 2.0, #Translation Rate\n",
    "    \n",
    "    #RNA Degredation Parameters\n",
    "    (\"rna_degredation_mm\", \"kb\"): 500, #RNAase-RNA binding (Default for RNA Degredation)\n",
    "    (\"rna_degredation_mm\", \"ku\"): 1.0, #RNAase-RNA unbinding (Default for RNA Degredation)\n",
    "    (\"rna_degredation_mm\", \"kdeg\"): 2.5 #Degredation Rate (Default for RNA Degredation)\n",
    "}\n",
    "              \n",
    "#Create a RegulatedPromoter Object named \"P_reg\" with regulators \"activator\" and \"repressor\"\n",
    "P_reg = bcp.RegulatedPromoter(\"P_reg\", regulators=[\"activator\", \"repressor\"], leak=True)\n",
    "\n",
    "#Create a DNA Assembly with an instance of the RegulatedPromoter\n",
    "reg_rep_assembly = bcp.DNAassembly(name=\"reporter\", promoter=P_reg, rbs=\"BCD\")\n",
    "\n",
    "#Define the activator and repressor species as proteins\n",
    "activator = bcp.Protein(\"activator\")\n",
    "repressor = bcp.Protein(\"repressor\")\n",
    "\n",
    "#Create a Mixture\n",
    "components = [reg_rep_assembly, activator, repressor]\n",
    "myMixture = bcp.BasicExtract(name=\"txtl\", parameters=parameters, components=components, parameter_warnings=True)\n",
    "\n",
    "myCRN = myMixture.compile_crn()\n",
    "print(\"\\n\"+repr(myCRN))\n",
    "\n",
    "#Simulate with BioSCRAPE, if installed\n",
    "try:\n",
    "    import pylab as plt\n",
    "    time = np.arange(0, 20, .01)\n",
    "    #Simulate the CRN with no activator or repressor\n",
    "    x0 = {\"protein_activator\":0, \"protein_repressor\":0, \"dna_reporter\":10, \"protein_Ribo\":100, \"protein_RNAP\":20, \"protein_RNAase\":10}\n",
    "    R_const = myCRN.simulate_with_bioscrape(time, stochastic = False, initial_condition_dict = x0)\n",
    "\n",
    "    #Simulate teh CRN with just repressor\n",
    "    x0 = {\"protein_activator\":0, \"protein_repressor\":50, \"dna_reporter\":10, \"protein_Ribo\":100, \"protein_RNAP\":20, \"protein_RNAase\":10}\n",
    "    R_repressed = myCRN.simulate_with_bioscrape(time, stochastic = False, initial_condition_dict = x0)\n",
    "\n",
    "    #Simulate the CRN with just activator\n",
    "    x0 = {\"protein_activator\":50, \"protein_repressor\":0, \"dna_reporter\":10, \"protein_Ribo\":100, \"protein_RNAP\":20, \"protein_RNAase\":10}\n",
    "    R_active = myCRN.simulate_with_bioscrape(time, stochastic = False, initial_condition_dict = x0)\n",
    "\n",
    "    plt.figure()\n",
    "    plt.plot(time, R_const[\"protein_reporter\"], label = \"Unregulated Expression\")\n",
    "    plt.plot(time, R_repressed[\"protein_reporter\"], label = \"Repressed Expression\")\n",
    "    plt.plot(time, R_active[\"protein_reporter\"], label = \"Activated Expression\")\n",
    "    plt.legend()\n",
    "    plt.xlabel(\"time\")\n",
    "    plt.ylabel(\"[Reporter]\")\n",
    "    plt.title(\"Regulated Promoter Example\")\n",
    "    plt.show()\n",
    "except ModuleNotFoundError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example 2: dCas9 \n",
    "In this example, dCas9 along with guide RNAs are used as regulator species for the RegulatedPromoter class from the previous example. In particular, a dCas9-gRNA complex is used as a repressor. We will also change the default binding mechanism from cooperative binding to reversible bimolecular binding as an example of how to change mechanisms. \n",
    "\n",
    "More specifically, this model includes constituitively expressed genes for a guide RNA. The dCas9:gRNA complex is then allowed to bind to a regulated promoter, lowering its affinity for RNAP thereby downregulating transcription. This model also includes loading effects on polymerases, RNAases, and Ribosomes - sensitiviy to all these parameters demonstrates some of the challenges of getting complex biological circuits to work well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "__init__() got an unexpected keyword argument 'material_type'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-58af43e0f05a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mdCas\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdCas_assembly\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;31m# Create a Guide RNA Guide1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0mgRNA\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbcp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mguideRNA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"gRNA\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdCas\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdCas\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0;31m# Create an assembly to express the guide RNA\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0mgRNA_assembly\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDNAassembly\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"gRNA\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtranscript\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgRNA\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpromoter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"P\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrbs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/sclamons/Downloads/biocrnpyler_local/biocrnpyler/d_cas9.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, guide_name, dCas9, **keywords)\u001b[0m\n\u001b[1;32m     22\u001b[0m             \u001b[0;34m\"dCas9_binding\"\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m                 Reversible_Bimolecular_Binding(name = \"dCas9_binding\",\n\u001b[0;32m---> 24\u001b[0;31m                                         material_type = \"bimolecular binding\")\n\u001b[0m\u001b[1;32m     25\u001b[0m         }\n\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: __init__() got an unexpected keyword argument 'material_type'"
     ]
    }
   ],
   "source": [
    "#Create some default parameters to be lazy. Most parameters will just use these\n",
    "parameters = {\n",
    "    \"ktx\":1.0, \"ktl\":3.0, \"kdeg\":.5, \"kb\":100, \"ku\":50, \"cooperativity\":1\n",
    "}\n",
    "\n",
    "# Create an assembly to express dCas9\n",
    "dCas_assembly = DNAassembly(\"dCas9\", promoter=\"P\", rbs=\"BCD\")\n",
    "# Get the dCas Species. Note: This could also be defined above and passed into the assembly as protein=...\n",
    "dCas = dCas_assembly.protein\n",
    "# Create a Guide RNA Guide1\n",
    "gRNA = bcp.guideRNA(\"gRNA\", dCas=dCas)\n",
    "# Create an assembly to express the guide RNA\n",
    "gRNA_assembly = DNAassembly(\"gRNA\", transcript=gRNA, promoter=\"P\", rbs=None)\n",
    "\n",
    "# Get the guideRNA:dCas9 complex\n",
    "repressor = bcp.gRNA.get_dCasComplex()\n",
    "\n",
    "# Create a Promoter regulated by the repressor\n",
    "P_reg = RegulatedPromoter(\"P_regulated\", regulators=[repressor], leak=True)\n",
    "P_reg.default_mechanisms['binding'] = Reversible_Bimolecular_Binding(\"dCas9_dna_binding\")\n",
    "#Use the custom mechanism name to set the parameters for all dCas9_dna_binding\n",
    "parameters[(\"dCas9_dna_binding\", \"ku\")] = .01\n",
    "parameters[(\"dCas9_dna_binding\", \"kb\")] = 1000\n",
    "\n",
    "\n",
    "#Set parameters of the repressed promoter\n",
    "parameters[(\"transcription\", P_reg.name+'_'+repressor.name, \"ku\")] = 1000\n",
    "parameters[(\"transcription\", P_reg.name+'_'+repressor.name, \"ktx\")] = .1\n",
    "#Set Parameters of the unrepressed promoter\n",
    "parameters[(\"transcription\", P_reg.name, \"ku\")] = .1\n",
    "parameters[(\"transcription\", P_reg.name, \"ktx\")] = 2.0\n",
    "\n",
    "# Create an assembly with the regulated promoter\n",
    "reg_assembly = bcp.DNAassembly(name=\"reporter\", promoter=P_reg, rbs=\"BCD\")\n",
    "\n",
    "# Create a list of components to add to the mixture (these could also be added one-by-one with Mixture.add_component(...)\n",
    "components = [dCas_assembly, gRNA_assembly, gRNA, reg_assembly]\n",
    "\n",
    "# Create a BasicExtract Mixture\n",
    "reaction_mix = bcp.BasicExtract(\"txtl\", components=components, parameters=parameters, parameter_warnings = False)\n",
    "\n",
    "# Compile a CRN\n",
    "CRN = reaction_mix.compile_crn()\n",
    "print(repr(CRN))\n",
    "\n",
    "print(\"Simulating with BioSCRAPE\")\n",
    "\n",
    "x0_no_dcas = {repr(dCas_assembly.get_specie()):0, \"dna_gRNA\":1, repr(reg_assembly.dna):5, \"protein_RNAP\":10, \"protein_RNAase\":10, \"protein_Ribo\":50}\n",
    "x0_with_dcas = {repr(dCas_assembly.get_specie()):1, \"dna_gRNA\":1, repr(reg_assembly.dna):5, \"protein_RNAP\":10, \"protein_RNAase\":10, \"protein_Ribo\":50}\n",
    "\n",
    "\n",
    "#Simulate with BioSCRAPE if installed\n",
    "try:\n",
    "    print(\"Simulating\")\n",
    "    import numpy as np\n",
    "    import pylab as plt\n",
    "    timepoints = np.arange(0, 20, .01)\n",
    "\n",
    "    stochastic = False\n",
    "    sim_no_cas = CRN.simulate_with_bioscrape(timepoints, x0_no_dcas, stochastic = stochastic)\n",
    "    sim_with_cas = CRN.simulate_with_bioscrape(timepoints, x0_with_dcas, stochastic = stochastic)\n",
    "\n",
    "\n",
    "    plt.figure(figsize = (10, 10))\n",
    "    plt.subplot(121)\n",
    "    plt.title(\"RNAs\")\n",
    "    plt.plot(timepoints, sim_no_cas[\"rna_gRNA\"], color = \"cyan\", label = \"gRNA: No dCas9\")\n",
    "    plt.plot(timepoints, sim_no_cas[\"rna_dCas9\"], color = \"blue\", label = \"dCas9 RNA: No dCas9\")\n",
    "    plt.plot(timepoints, sim_no_cas[\"rna_reporter\"], color = \"green\", label = \"Reporter RNA: No dCas9\")\n",
    "\n",
    "    plt.plot(timepoints, sim_with_cas[\"rna_gRNA\"], \":\", color = \"cyan\", label = \"gRNA: with dCas9\")\n",
    "    plt.plot(timepoints, sim_with_cas[\"rna_dCas9\"], \":\", color = \"blue\", label = \"dCas9 RNA: with dCas9\")\n",
    "    plt.plot(timepoints, sim_with_cas[\"rna_reporter\"], \":\", color = \"green\", label = \"Reporter RNA: with dCas9\")\n",
    "\n",
    "    plt.legend()\n",
    "    \n",
    "    plt.subplot(122)\n",
    "    plt.title(\"proteins\")\n",
    "    plt.plot(timepoints, sim_no_cas[\"protein_dCas9\"], color = \"blue\", label = \"dCas9 protein: No dCas9\")\n",
    "    plt.plot(timepoints, sim_no_cas[\"protein_reporter\"], color = \"green\", label = \"Reporter protein: No dCas9\")\n",
    "\n",
    "    plt.plot(timepoints, sim_with_cas[\"protein_dCas9\"], \":\", color = \"blue\", label = \"dCas9 protein: with dCas9\")\n",
    "    plt.plot(timepoints, sim_with_cas[\"protein_reporter\"], \":\", color = \"green\", label = \"Reporter protein: with dCas9\")\n",
    "    plt.legend()\n",
    "\n",
    "    plt.plot()\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "except ModuleNotFoundError:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dilution with Global Mechanisms\n",
    "Global mechanisms provide a framework to have certain reactions work on all, or almost all, species in a CRN. They are to be used with caution: depending on the order global mechanisms are applied, different resulting CRNs are possible. In fact, the recommendation is to only use global mechanisms for degredation by dilution. In the following example, a model will be set up where global mechanisms cause degredation by dilution on all species which are not labeled as \"genomic\" or \"machinery\".\n",
    "\n",
    "Under the hood, global mechanisms work just like normal mechanisms - they take a set of species and compile them into CRNs. Global mechanisms are called at the end of the compilation, so they apply to all species generated by all local mechanisms. In order to make global mechanisms more selective, a filter_dict keywork can be used. The filter_dict (str:True/False) takes some str such as \"dna\" or \"genomic\" which is either a specie type, name, or attribute. The mechanism is then applied or not based upon the value of that attribute in the filter_dict. For exampe filter_dict = {\"dna\":False} would not apply its mechanism to any \"dna\" species. For species with no attributes in the filter_dict, a global mechanism defaults to its default_on keyword, which can be True or False.\n",
    "\n",
    "In the below example, we will show how filters and attributes work on constuitively expressed genes using dilution as the prototypical example. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species = protein_RNAP_machinery, protein_Ribo_machinery, protein_RNAase_machinery, complex_rna_G1_protein_RNAase_machinery, complex_dna_G1_protein_RNAP_machinery, complex_rna_G1_protein_Ribo_machinery, protein_G1, dna_G1, rna_G1, protein_G2, dna_G2_genomic, complex_rna_G2_protein_RNAase_machinery, complex_dna_G2_protein_RNAP_machinery_genomic, complex_rna_G2_protein_Ribo_machinery, rna_G2\n",
      "Reactions = [\n",
      "\tdna_G1 + protein_RNAP_machinery <--> complex_dna_G1_protein_RNAP_machinery        massaction: k_f=100,\tk_r=20\n",
      "\tcomplex_dna_G1_protein_RNAP_machinery --> dna_G1 + rna_G1 + protein_RNAP_machinery        massaction: k_f=self.k\n",
      "\trna_G1 + protein_Ribo_machinery <--> complex_rna_G1_protein_Ribo_machinery        massaction: k_f=100,\tk_r=20\n",
      "\tcomplex_rna_G1_protein_Ribo_machinery --> rna_G1 + protein_G1 + protein_Ribo_machinery        massaction: k_f=self.k\n",
      "\trna_G1 + protein_RNAase_machinery <--> complex_rna_G1_protein_RNAase_machinery        massaction: k_f=100,\tk_r=20\n",
      "\tcomplex_rna_G1_protein_RNAase_machinery --> protein_RNAase_machinery        massaction: k_f=self.k\n",
      "\tdna_G2_genomic + protein_RNAP_machinery <--> complex_dna_G2_protein_RNAP_machinery_genomic        massaction: k_f=100,\tk_r=20\n",
      "\tcomplex_dna_G2_protein_RNAP_machinery_genomic --> dna_G2_genomic + rna_G2 + protein_RNAP_machinery        massaction: k_f=self.k\n",
      "\trna_G2 + protein_Ribo_machinery <--> complex_rna_G2_protein_Ribo_machinery        massaction: k_f=100,\tk_r=20\n",
      "\tcomplex_rna_G2_protein_Ribo_machinery --> rna_G2 + protein_G2 + protein_Ribo_machinery        massaction: k_f=self.k\n",
      "\trna_G2 + protein_RNAase_machinery <--> complex_rna_G2_protein_RNAase_machinery        massaction: k_f=100,\tk_r=20\n",
      "\tcomplex_rna_G2_protein_RNAase_machinery --> protein_RNAase_machinery        massaction: k_f=self.k\n",
      "\tprotein_G1 -->         massaction: k_f=self.k\n",
      "\tdna_G1 -->         massaction: k_f=self.k\n",
      "\trna_G1 -->         massaction: k_f=self.k\n",
      "\tprotein_G2 -->         massaction: k_f=self.k\n",
      "\trna_G2 -->         massaction: k_f=self.k\n",
      "]\n",
      "Simulating with BioSCRAPE\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'py_simulate_model'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-9e66f9e203a6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[0mx0_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmyMixture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mribosome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_species\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmyMixture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrnap\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_species\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmyMixture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRNAase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_species\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mA_dna\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdna\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mA_genome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdna\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 42\u001b[0;31m     \u001b[0mfull_result_sto\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmyCRN\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimulate_with_bioscrape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimepoints\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx0_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstochastic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     43\u001b[0m     \u001b[0mfull_result_det\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmyCRN\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimulate_with_bioscrape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimepoints\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx0_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstochastic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/sclamons/Downloads/biocrnpyler_local/biocrnpyler/chemical_reaction_network.py\u001b[0m in \u001b[0;36msimulate_with_bioscrape\u001b[0;34m(self, timepoints, initial_condition_dict, stochastic, return_dataframe, safe)\u001b[0m\n\u001b[1;32m    567\u001b[0m                                 \u001b[0mstochastic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_dataframe\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    568\u001b[0m                                 safe = True):\n\u001b[0;32m--> 569\u001b[0;31m         \u001b[0;32mfrom\u001b[0m \u001b[0mbioscrape\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimulator\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpy_simulate_model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    570\u001b[0m         \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_bioscrape_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    571\u001b[0m         \u001b[0mm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_species\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minitial_condition_dict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'py_simulate_model'"
     ]
    }
   ],
   "source": [
    "from biocrnpyler import *\n",
    "\n",
    "#We will only use default parameters in this model, for simplicity.\n",
    "kb, ku, ktx, ktl, kdeg, kdil = 100, 20, 3, 2, .5, .5\n",
    "parameters = {\"kb\":kb, \"ku\":ku, \"ktx\":ktx, \"ktl\":ktl, \"kdeg\": kdeg, \"kdil\":kdil}\n",
    "\n",
    "#Creates a global dilution mechanism that acts on all species generated except for\n",
    "# those with the type or attribute \"genome\" or \"machinery\"\n",
    "dilution_mechanism = Dilution(filter_dict = {\"genomic\":False, \"machinery\":False}, default_on = True)\n",
    "\n",
    "#Add this mechanism to a dictionary which is passed into the Mixture txtl.BasicExtract\n",
    "global_mechanisms = {\"dilution\":dilution_mechanism}\n",
    "myMixture = BasicExtract(name = \"txtl\", parameters = parameters, global_mechanisms = global_mechanisms, parameter_warnings= False)\n",
    "\n",
    "#Add machinery attributes to species I want constiutively expressed at the dilution rate\n",
    "myMixture.rnap.add_attribute(\"machinery\")\n",
    "myMixture.RNAase.add_attribute(\"machinery\")\n",
    "myMixture.ribosome.add_attribute(\"machinery\")\n",
    "\n",
    "#Creates a dna assembly. This assembly is type \"dna\" so it will be degraded\n",
    "A_dna = DNAassembly(name = \"G1\", promoter = \"pBest\", rbs = \"BCD2\")\n",
    "\n",
    "#Create another dna assembly but set its internal specie's attributes to contain \"genomic\" so it will not be degraded\n",
    "#Note: this only protects the dna_G2 species encoded by this assembly as well as complex species (eg rnap:DNA) which inherit their subspecies attributes.\n",
    "A_genome = DNAassembly(name = \"G2\", promoter = \"pBest\", rbs = \"BCD2\", attributes = [\"genomic\"])\n",
    "\n",
    "\n",
    "myMixture.add_components(A_dna)\n",
    "myMixture.add_components(A_genome)\n",
    "myCRN = myMixture.compile_crn()\n",
    "print(repr(myCRN))\n",
    "\n",
    "\n",
    "try:\n",
    "    print(\"Simulating with BioSCRAPE\")\n",
    "    import numpy as np\n",
    "    import pylab as plt\n",
    "    timepoints = np.arange(0, 50, .1)\n",
    "\n",
    "    x0_dict = {repr(myMixture.ribosome.get_species()):100, repr(myMixture.rnap.get_species()):20, repr(myMixture.RNAase.get_species()):10, repr(A_dna.dna):20, repr(A_genome.dna):20}\n",
    "\n",
    "    full_result_sto = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = True)\n",
    "    full_result_det = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = False)\n",
    "\n",
    "    #chemical_reaction_network.get_all_species_containing is a useful shortcut to get lists of species\n",
    "    tot_A_dna_det = np.sum(full_result_det[myCRN.get_all_species_containing(A_dna.dna, return_as_strings=True)], 1)\n",
    "    tot_A_genome_det = np.sum(full_result_det[myCRN.get_all_species_containing(A_genome.dna, return_as_strings=True)], 1)\n",
    "    tot_A_dna_sto = np.sum(full_result_sto[myCRN.get_all_species_containing(A_dna.dna, return_as_strings=True)], 1)\n",
    "    tot_A_genome_sto = np.sum(full_result_sto[myCRN.get_all_species_containing(A_genome.dna, return_as_strings=True)], 1)\n",
    "\n",
    "    tot_A_dna_rna_det = np.sum(full_result_det[myCRN.get_all_species_containing(A_dna.transcript, return_as_strings=True)], 1)\n",
    "    tot_A_genome_rna_det = np.sum(full_result_det[myCRN.get_all_species_containing(A_genome.protein, return_as_strings=True)], 1)\n",
    "    tot_A_dna__rna_sto = np.sum(full_result_sto[myCRN.get_all_species_containing(A_dna.transcript, return_as_strings=True)], 1)\n",
    "    tot_A_genome_rna_sto = np.sum(full_result_sto[myCRN.get_all_species_containing(A_genome.protein, return_as_strings=True)], 1)\n",
    "\n",
    "\n",
    "    plt.figure(figsize = (14, 10))\n",
    "    plt.subplot(131)\n",
    "    plt.plot(timepoints, tot_A_dna_det, color = \"blue\", label = \"Non-Genomic DNA (deterministic)\")\n",
    "    plt.plot(timepoints, tot_A_genome_det, color = \"cyan\", label = \"Genomic DNA (deterministic)\")\n",
    "    plt.plot(timepoints, tot_A_dna_sto, \":\", color = \"blue\", label = \"Non-Genomic DNA (stochastic)\")\n",
    "    plt.plot(timepoints, tot_A_genome_sto, \":\", color = \"cyan\", label = \"Genomic DNA (stochastic)\")\n",
    "    plt.legend()\n",
    "    plt.xlabel(\"Time\")\n",
    "    plt.title(\"DNA\")\n",
    "    plt.ylabel(\"Concentration or Count\")\n",
    "\n",
    "    plt.subplot(132)\n",
    "    plt.plot(timepoints, tot_A_dna_rna_det, color = \"blue\", label = \"Non-Genomic RNA (deterministic)\")\n",
    "    plt.plot(timepoints, tot_A_genome_rna_det, color = \"cyan\", label = \"Genomic RNA (deterministic)\")\n",
    "    plt.plot(timepoints, tot_A_dna__rna_sto, \":\", color = \"blue\", label = \"Non-Genomic RNA (stochastic)\")\n",
    "    plt.plot(timepoints, tot_A_genome_rna_sto, \":\", color = \"cyan\", label = \"Genomic RNA (stochastic)\")\n",
    "    plt.legend()\n",
    "    plt.xlabel(\"Time\")\n",
    "    plt.title(\"RNA\")\n",
    "    plt.ylabel(\"Concentration / Count\")\n",
    "\n",
    "    plt.subplot(133)\n",
    "    plt.plot(timepoints, full_result_det[str(A_dna.protein)], color = \"blue\", label = \"Non-Genomic Protein (deterministic)\")\n",
    "    plt.plot(timepoints, full_result_det[str(A_genome.protein)], color = \"cyan\", label = \"Genomic Protein (deterministic)\")\n",
    "    plt.plot(timepoints, full_result_sto[str(A_dna.protein)], \":\", color = \"blue\", label = \"Non-Genomic Protein (stochastic)\")\n",
    "    plt.plot(timepoints, full_result_sto[str(A_genome.protein)], \":\", color = \"cyan\", label = \"Genomic Protein (stochastic)\")\n",
    "    plt.legend()\n",
    "    plt.title(\"Protein\")\n",
    "    plt.xlabel(\"Time\")\n",
    "    plt.ylabel(\"Concentration / Count\")\n",
    "\n",
    "    plt.show()\n",
    "except ModuleNotFoundError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
